src/main/java/seedu/address/ui/CommandHistory.java

```diff
@@ -0,0 +1,93 @@
+package seedu.address.ui;
+
+import static java.util.Objects.requireNonNull;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.NoSuchElementException;
+import java.util.Optional;
+
+/**
+ * Represents a stack of commands executed by the user.
+ * Supports caching the current command before starting to traverse the stack from the top.
+ */
+public class CommandHistory {
```

Comment:
I don't really quite get the purpose of the cache.

Reply:
The cache is used in the following scenario:
1. The user has some previous commands already saved in the history.
2. The user partially types a command string, but decides to refer to a previous command.
3. The user recalls the previous command for reference but wants to return to the partially typed command.
4. The user can retrieve his partially typed command string from the cache.

I've also included some logging functionality at the info and fine levels.

Reply:
Understood the cache. Logger lgtm.